src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs

```diff
@@ -835,6 +781,22 @@ void addTemp(BoundDagEvaluation e, TypeSymbol t, int index = 0)
                 Debug.Assert(outputSlot > 0);
                 addToTempMap(output, outputSlot, type.Type);
             }
+
+            static TypeWithAnnotations getIndexerOutputType(TypeSymbol inputType, BoundExpression e, bool isSlice)
+            {
+                return e switch
+                {
+                    BoundIndexerAccess indexerAccess => AsMemberOfType(inputType, indexerAccess.Indexer).GetTypeOrReturnType(),
+                    BoundCall call => AsMemberOfType(inputType, call.Method).GetTypeOrReturnType(),
+
+                    BoundArrayAccess arrayAccess => isSlice
+                        ? TypeWithAnnotations.Create(isNullableEnabled: true, inputType, isAnnotated: false)
+                        : ((ArrayTypeSymbol)inputType).ElementTypeWithAnnotations,
+
+                    BoundIndexOrRangePatternIndexerAccess implicitIndexerAccess => getIndexerOutputType(inputType, implicitIndexerAccess.IndexerAccess, isSlice),
+                    _ => throw ExceptionUtilities.UnexpectedValue(e.Kind)
```

Comment:
>throw ExceptionUtilities.UnexpectedValue(e.Kind) [](http://example.com/codeflow?start=25&length=48)

It looks like an assert in binding mentions ```BoundDynamicIndexerAccess``` and ```BoundBadExpression```.

Reply:
Tested with `dynamic`. We don't construct a full DAG in such error case, so no BounDynamicIndexerAccess makes it into the DAG.
```
State 0
    1. [[_]] AND(?DagNonNullTest(t0), ?DagTypeTest(t0 is ?))
    Test: ?DagTypeTest(t0 is ?)
    TrueBranch: 1
    FalseBranch: 2
*State 1
    1. [[_]] MATCH
*State 2 FAIL
```

That's because of
```
    if (list.HasErrors)
            {
                tests.Add(new Tests.One(new BoundDagTypeTest(list.Syntax, ErrorType(), input, hasErrors: true)));
            }
```